### DISCUSSION TITLE
Is assert best practise?

### DISCUSSION THREAD

**Discussion by CTimmerman** [pytest-dev/pytest#9932] [Q&A] [answered]
https://docs.pytest.org/en/6.2.x/assert.html says `assert` is valid:
```
# content of test_assert1.py
def f(): return 3

def test_function(): assert f() == 4
```

However, Bandit warns that `Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.`

So should users do this instead?
```
def test_function():
    if f() != 4: raise RuntimeError()
```

**Comment by The-Compiler** [comment, marked as answer]
The warning you get from bandit applies (in certain conditions) to your application/library code - but it does not apply to test code. There is no point in running your testsuite with assertions disabled.

**Reply by CTimmerman** [reply]
The point of optimizing is to increase speed, which some huge testsuites might want, but if optimizing always removes assertions that's a thing to note.

**Reply by nicoddemus** [reply]
I think it is already noted in the docs, and you also get a warning from pytest if you disable assertions.
